ckanext/qdes/logic/action/get.py

```diff
@@ -0,0 +1,341 @@
+                'Point of contact - email': _qdes_extract_point_of_contact(package.get('contact_point', None), 'Email'),
+                'List of recommended fields without values': ', '.join(missing_values),
+                'Organisation name': package.get('organization').get('title', ''),
+            })
+
+            # Check dataset resource metadata fields.
+            for resource in package.get('resources', []):
+                missing_values = _check_recommended_field_value(resource, dataset_resource_recommended_fields)
+                rows.append({
+                    'Dataset name': package.get('title', package.get('name', '')),
+                    'Link to dataset (URI)': url_for('dataset.read', id=package.get('id'), _external=True),
+                    'Resource name': resource.get('name', ''),
+                    'Link to resource': url_for('resource.read',
+                                                resource_id=resource.get('id'),
+                                                id=package.get('id'),
+                                                package_type=package.get('type'),
+                                                _external=True
+                                                ),
+                    'Dataset creator': package.get('contact_creator', ''),
+                    'Point of contact - name':
+                        _qdes_extract_point_of_contact(package.get('contact_point', None), 'Name'),
+                    'Point of contact - email':
+                        _qdes_extract_point_of_contact(package.get('contact_point', None), 'Email'),
+                    'List of recommended fields without values': ', '.join(missing_values),
+                    'Organisation name': package.get('organization').get('title', ''),
+                })
+
+    return rows
+
+
+def _get_uri_validated_fields(scheme, field_group):
+    uri_validated_fields = []
+    for field in scheme.get(field_group, []):
+        validator = field.get('validators', '').split()
+        if 'qdes_uri_validator' in validator:
+            uri_validated_fields.append(field)
+
+    return uri_validated_fields
+
+
+def qdes_datasets_with_invalid_urls(context, config):
+    u"""
+    List of all datasets with broken links to resources.
+    """
+    # Get list of invalid uris.
+    uris = Session.query(InvalidUri).all()
+    invalid_uris = [uri.as_dict() for uri in uris]
+
+    # Build package list.
+    entities = {}
+    for invalid_uri in invalid_uris:
+        if invalid_uri.get('entity_id') in entities:
+            entities[invalid_uri.get('entity_id')].get('fields').append(invalid_uri.get('field'))
+        else:
+            entities[invalid_uri.get('entity_id')] = {}
+            entities[invalid_uri.get('entity_id')].update({
+                'type': invalid_uri.get('entity_type'),
+                'fields': [invalid_uri.get('field')],
+            })
```

Comment:
Hi @awset - this block (from `# Get list of invalid uris.` down to `# Build rows.`) can be simplified for example:
```
    # Get list of invalid uris.
    invalid_uris = Session.query(InvalidUri).all()

    # Build package list.
    entities = {}
    for uri in invalid_uris:
        if uri.entity_id in entities:
            entities[uri.entity_id]['fields'].append(uri.field)
        else:
            entities[uri.entity_id] = {
                'type': uri.entity_type,
                'fields': [uri.field],
            }
```
...makes the code easier to read & maintain. It could possibly be extracted to a helper function too, to reduce the function size for maintainability down the track

Reply:
I have moved it to helper function